Minkowski multiplication for semigroups, monoids, and commutative monoids (#1309)
UniMath/agda-unimath: src/group-theory/minkowski-multiplication-commutative-monoids.lagda.md

> # Minkowski multiplication of commutative monoid subtypes

Reviewer: ```suggestion
# Minkowski multiplication on subsets of a commutative monoid
```

Author: I promise, you can just tell me to fix all of these, and I'll go find them myself -- you don't have to find them all and fix them yourself!

```agda
module group-theory.minkowski-multiplication-commutative-monoids where
```

<details><summary>Imports</summary>

```agda
open import foundation.dependent-pair-types
open import foundation.existential-quantification
open import foundation.identity-types
open import foundation.inhabited-subtypes
open import foundation.powersets
open import foundation.subtypes
open import foundation.universe-levels

open import group-theory.commutative-monoids
open import group-theory.minkowski-multiplication-monoids
open import group-theory.monoids
open import group-theory.subsets-commutative-monoids
```

</details>

## Idea

> For two [subsets](group-theory.subsets-commutative-monoids.md) `A`, `B` of a
> [commutative monoid](group-theory.commutative-monoids.md) `M`, the Minkowski
> multiplication of `A` and `B` is the set of elements that can be formed by
> multiplying an element of `A` and an element of `B`. (This is more usually
> referred to as a Minkowski sum, but as the operation on commutative monoids is
> referred to as `mul`, we use multiplicative terminology.)

Reviewer: ```suggestion
Given two [subsets](group-theory.subsets-commutative-monoids.md) `A` and `B` of a
[commutative monoid](group-theory.commutative-monoids.md) `M`, the {{#concept "Minkowski
multiplication" Disambiguation="on subsets of a commutative monoid" Agda=minkowski-mul-Commutative-Monoid}} of `A` and `B` is the [set](foundation-core.sets.md) of elements that can be formed by
multiplying an element of `A` and an element of `B`.
```

Reviewer 2: I'm viewing this on my phone, but it looks like your suggestion has a linebreak in a concept, which shouldn't be there. Also Minkowski addition has a wikidata id that we should link to by adding the WDID and WD properties

Reviewer: Thanks for pointing it out

Reviewer: fixed

Author: Aha.  I hadn't realized what `WDID` and `WD` stand for, or I would've figured out how to add them myself.  Is there some documentation somewhere saying that that I missed?

Reviewer: It doesn't look like this change got implemented.

Reviewer 2: > Is there some documentation somewhere saying that that I missed?

Nope, the `#concept` macro is still undocumented. It's original purpose was to tag our formalization with wikidata identifiers, so that this other project could create a hub presenting formalizations of one concept in different libraries. It was pretty underspecified, and the other project is now down, but we kept the annotations for organizational purposes. But no one ever did any requirements gathering, I just implemented whatever came to mind and wasn't too time consuming, expecting us to one day sit down and specify what we want it to do

## Definition

```agda
module _
  {l1 l2 l3 : Level}
  (M : Commutative-Monoid l1)
  (A : subset-Commutative-Monoid l2 M)
  (B : subset-Commutative-Monoid l3 M)
  where

  minkowski-mul-Commutative-Monoid :
    subset-Commutative-Monoid (l1 ⊔ l2 ⊔ l3) M
  minkowski-mul-Commutative-Monoid =
    minkowski-mul-Monoid (monoid-Commutative-Monoid M) A B
```

## Properties

> ### Minkowski multiplication on a commutative monoid is commutative

Reviewer: ```suggestion
### Minkowski multiplication on subsets of a commutative monoid is commutative
```

```agda
module _
  {l1 l2 l3 : Level}
  (M : Commutative-Monoid l1)
  (A : subset-Commutative-Monoid l2 M)
  (B : subset-Commutative-Monoid l3 M)
  where

  commutative-minkowski-mul-leq-Commutative-Monoid :
    minkowski-mul-Commutative-Monoid M A B ⊆
    minkowski-mul-Commutative-Monoid M B A
  commutative-minkowski-mul-leq-Commutative-Monoid x =
    elim-exists
      (minkowski-mul-Commutative-Monoid M B A x)
      ( λ (a , b) (a∈A , b∈B , x=ab) →
        intro-exists
          ( b , a)

>           ( b∈B , a∈A , x=ab ∙ commutative-mul-Commutative-Monoid M a b))

Reviewer: This looks like a `map-exists` to me 👀👀

Author: FWIW, my experimentation has found that the leading argument of `map-exists` tends to mean that it doesn't really simplify much.  I'll do it here, though.

Author: I've done it, but it honestly looks to me like it's made things substantially more complicated.

Reviewer: I agree with you. Feel free to revert.

module _
  {l1 l2 l3 : Level}
  (M : Commutative-Monoid l1)
  (A : subset-Commutative-Monoid l2 M)
  (B : subset-Commutative-Monoid l3 M)
  where

  commutative-minkowski-mul-Commutative-Monoid :
    minkowski-mul-Commutative-Monoid M A B ＝
    minkowski-mul-Commutative-Monoid M B A
  commutative-minkowski-mul-Commutative-Monoid =
    antisymmetric-sim-subtype
      ( minkowski-mul-Commutative-Monoid M A B)
      ( minkowski-mul-Commutative-Monoid M B A)
      ( commutative-minkowski-mul-leq-Commutative-Monoid M A B ,
        commutative-minkowski-mul-leq-Commutative-Monoid M B A)
```

> ### Minkowski multiplication on a commutative monoid is a commutative monoid

Reviewer: ```suggestion
### Minkowski multiplication on subsets of a commutative monoid is a commutative monoid
```

```agda
module _
  {l : Level}
  (M : Commutative-Monoid l)
  where

  commutative-monoid-minkowski-mul-Commutative-Monoid :
    Commutative-Monoid (lsuc l)
  pr1 commutative-monoid-minkowski-mul-Commutative-Monoid =
    monoid-minkowski-mul-Monoid (monoid-Commutative-Monoid M)
  pr2 commutative-monoid-minkowski-mul-Commutative-Monoid =
    commutative-minkowski-mul-Commutative-Monoid M
```

> ### The Minkowski multiplication of two inhabited subsets of a commutative monoid is inhabited

Reviewer: ```suggestion
### The Minkowski multiplication of two inhabited subsets is inhabited
```

```agda
module _
  {l1 : Level}
  (M : Commutative-Monoid l1)
  where

  minkowski-mul-inhabited-is-inhabited-Commutative-Monoid :
    {l2 l3 : Level} →
    (A : subset-Commutative-Monoid l2 M) →
    (B : subset-Commutative-Monoid l3 M) →
    is-inhabited-subtype A →
    is-inhabited-subtype B →
    is-inhabited-subtype (minkowski-mul-Commutative-Monoid M A B)
  minkowski-mul-inhabited-is-inhabited-Commutative-Monoid =
    minkowski-mul-inhabited-is-inhabited-Monoid (monoid-Commutative-Monoid M)
```

> ### Containment is preserved by Minkowski multiplication of monoid subsets

Reviewer: ```suggestion
### Containment of subsets is preserved by Minkowski multiplication
```

```agda
module _
  {l1 l2 l3 l4 : Level}
  (M : Commutative-Monoid l1)
  (B : subset-Commutative-Monoid l2 M)
  (A : subset-Commutative-Monoid l3 M)
  (A' : subset-Commutative-Monoid l4 M)
  where

  preserves-leq-left-minkowski-mul-Commutative-Monoid :
    A ⊆ A' →
    minkowski-mul-Commutative-Monoid M A B ⊆
    minkowski-mul-Commutative-Monoid M A' B
  preserves-leq-left-minkowski-mul-Commutative-Monoid =
    preserves-leq-left-minkowski-mul-Monoid (monoid-Commutative-Monoid M) B A A'

  preserves-leq-right-minkowski-mul-Commutative-Monoid :
    A ⊆ A' →
    minkowski-mul-Commutative-Monoid M B A ⊆
    minkowski-mul-Commutative-Monoid M B A'
  preserves-leq-right-minkowski-mul-Commutative-Monoid =
    preserves-leq-right-minkowski-mul-Monoid
      ( monoid-Commutative-Monoid M)
      ( B)
      ( A)
      ( A')
```

> ### Similarity is preserved by Minkowski multiplication of monoid subsets

Reviewer: ```suggestion
### Similarity of subsets is preserved by Minkowski multiplication
```

open import group-theory.minkowski-multiplication-monoids
open import group-theory.monoids
open import group-theory.subsets-commutative-monoids

open import logic.functoriality-existential-quantification
```

</details>

## Idea

Given two [subsets](group-theory.subsets-commutative-monoids.md) `A` and `B` of
a [commutative monoid](group-theory.commutative-monoids.md) `M`, the
{{#concept "Minkowski multiplication" Disambiguation="on subsets of a commutative monoid" WD="Minkowski addition" WDID=Q1322294 Agda=minkowski-mul-Commutative-Monoid}}
of `A` and `B` is the [set](foundation-core.sets.md) of elements that can be
formed by multiplying an element of `A` and an element of `B`. This binary
operation defines a commutative monoid structure on the
[powerset](foundation.powersets.md) of `M`.

## Definition

```agda
module _
  {l1 l2 l3 : Level}
  (M : Commutative-Monoid l1)
  (A : subset-Commutative-Monoid l2 M)
  (B : subset-Commutative-Monoid l3 M)
  where

  minkowski-mul-Commutative-Monoid :
    subset-Commutative-Monoid (l1 ⊔ l2 ⊔ l3) M
  minkowski-mul-Commutative-Monoid =
    minkowski-mul-Monoid (monoid-Commutative-Monoid M) A B
```

## Properties

> ### Minkowski multiplication of commutative monoid subsets is associative

Reviewer: ```suggestion
### Minkowski multiplication on subsets of a commutative monoid is associative
```

```agda
module _
  {l1 l2 l3 l4 : Level}
  (M : Commutative-Monoid l1)
  (A : subset-Commutative-Monoid l2 M)
  (B : subset-Commutative-Monoid l3 M)
  (C : subset-Commutative-Monoid l4 M)
  where

  associative-minkowski-mul-Commutative-Monoid :
    minkowski-mul-Commutative-Monoid
      ( M)
      ( minkowski-mul-Commutative-Monoid M A B)
      ( C) ＝
    minkowski-mul-Commutative-Monoid
      ( M)
      ( A)
      ( minkowski-mul-Commutative-Monoid M B C)
  associative-minkowski-mul-Commutative-Monoid =
    associative-minkowski-mul-Monoid (monoid-Commutative-Monoid M) A B C
```

> ### Minkowski multiplication of commutative monoid subsets is unital

Reviewer: ```suggestion
### Minkowski multiplication on subsets of a commutative monoid is unital
```
